### PR TITLE
[DependencyInjection] Improve ParameterNotFoundException when accessing a nested parameter

### PR DESCRIPTION
**Pull request by wouterj**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | n/a |

A common problem under beginners is to think that the dot notation is used to access nested arrays saved in parameters (_common here means someone asks help for this problem at least once a week on IRC_). Adding a little extra detail to the exception message and a working alternative should help pointing these people in the right direction before spending time debugging this.

It's quite late in the night over here, so the wording of the exception message probably isn't great. I'm happy to accept better suggestions 😃

### CODE REVIEW

**Comment by wouterj** [comment]
travis failure unrelated

File: src/Symfony/Component/DependencyInjection/Exception/ParameterNotFoundException.php [hunk 1]
```diff
@@ -30,12 +31,13 @@ class ParameterNotFoundException extends InvalidArgumentException
      * @param \Exception $previous     The previous exception
      * @param string[]   $alternatives Some parameter name alternatives
      */
-    public function __construct($key, $sourceId = null, $sourceKey = null, \Exception $previous = null, array $alternatives = array())
+    public function __construct($key, $sourceId = null, $sourceKey = null, \Exception $previous = null, array $alternatives = array(), $nonNestedAlternative = null)
```
Comment (kashike) [inline src/Symfony/Component/DependencyInjection/Exception/ParameterNotFoundException.php]: `$nonNestedAlternative` should be added to docblock, no?

Comment (wouterj) [inline src/Symfony/Component/DependencyInjection/Exception/ParameterNotFoundException.php]: fixed

File: src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php [hunk 2]
```diff
@@ -81,7 +81,23 @@ public function get($name)
                 }
             }
 
-            throw new ParameterNotFoundException($name, null, null, null, $alternatives);
+            $nonNestedAlternative = null;
+            if (!count($alternatives) && false !== strpos($name, '.')) {
+                $namePartsLength = array_map('strlen', explode('.', $name));
```
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php]: What about using implode ?

File: src/Symfony/Component/DependencyInjection/Exception/ParameterNotFoundException.php [hunk 3]
```diff
@@ -59,6 +62,8 @@ public function updateRepr()
                 $this->message .= ' Did you mean one of these: "';
             }
             $this->message .= implode('", "', $this->alternatives).'"?';
+        } elseif (null !== $this->nonNestedAlternative) {
+            $this->message .= ' You cannot access nested array items, do you want to inject "'.$this->nonNestedAlternative.'" instead?';
```
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Exception/ParameterNotFoundException.php]: What about using `Did you mean` to be consistent ?

**Comment by fabpot** [comment]
Thank you @wouterj.

**Pull request closed without merging**